### PR TITLE
KDESKTOP-980-Dehydrated-placeholders-exclusion-doesn-t-work

### PR DESCRIPTION
**Pull request by ClementKunz**
Sometimes the app detect an EDIT and try to upload a dehydrated placeholder. In those cases, the operation is ignored but the app continue to try to upload on each sync.
To avoid that now. we update the date of the dehydrated placeholder with the one in DB.

### CODE REVIEW

File: src/server/socketapi.cpp [hunk 1]
```diff
@@ -345,11 +349,9 @@ void SocketApi::command_RETRIEVE_FILE_STATUS(const QString &argument, SocketList
         return;
     }
 
-    if (status != KDC::SyncFileStatusUnknown) {
-        QString message =
-            buildMessage(QString("STATUS"), fileData._localPath, socketAPIString(status, isPlaceholder, isHydrated, progress));
-        listener->sendMessage(message);
-    }
+    QString message =
```
Comment (luc-guyot-infomaniak) [inline src/server/socketapi.cpp:352]:
```suggestion
    const QString message =
```

File: src/libsyncengine/propagation/executor/executorworker.cpp [hunk 2]
```diff
@@ -1073,6 +1074,34 @@ bool ExecutorWorker::checkLiteSyncInfoForEdit(SyncOpPtr syncOp, SyncPath &absolu
                                             << Utility::formatSyncPath(absolutePath).c_str() << L" (" << Utility::s2ws(id).c_str()
                                             << L")");
             ignoreItem = true;
+
+            // Update last modification date in order to avoid generating more EDIT operations.
```
Comment (luc-guyot-infomaniak) [inline src/libsyncengine/propagation/executor/executorworker.cpp:1078]: **Suggestion**: Encapsulate this addition within a method whose docstring would document the issue that is being fixed.

Are some software suspected to be responsible for those spurious EDIT operations?

Comment (ClementKunz) [reply src/libsyncengine/propagation/executor/executorworker.cpp]: Done.
No, we do not have clue yet on how those modification dates have changed.

File: src/libsyncengine/propagation/executor/executorworker.h [hunk 3]
```diff
@@ -55,7 +55,8 @@ class ExecutorWorker : public OperationProcessor {
 
         void handleEditOp(SyncOpPtr syncOp, std::shared_ptr<AbstractJob> &job, bool &hasError);
         bool generateEditJob(SyncOpPtr syncOp, std::shared_ptr<AbstractJob> &job);
-        bool checkLiteSyncInfoForEdit(SyncOpPtr syncOp, SyncPath &absolutePath, bool &ignoreItem,
+        bool fixModificationDate(SyncOpPtr syncOp, const SyncPath &absolutePath);
```
[the remark this answers is not in the document]
Comment (ClementKunz) [reply src/libsyncengine/propagation/executor/executorworker.h:66]: Done

File: src/libsyncengine/propagation/executor/executorworker.h [hunk 4]
```diff
@@ -55,7 +55,16 @@ class ExecutorWorker : public OperationProcessor {
 
         void handleEditOp(SyncOpPtr syncOp, std::shared_ptr<AbstractJob> &job, bool &hasError);
         bool generateEditJob(SyncOpPtr syncOp, std::shared_ptr<AbstractJob> &job);
-        bool checkLiteSyncInfoForEdit(SyncOpPtr syncOp, SyncPath &absolutePath, bool &ignoreItem,
+    
+        /**
+         * This method aims to fix the last modification date of a local file using the date stored in DB. This allow to fix wrong
```
Comment (luc-guyot-infomaniak) [inline src/libsyncengine/propagation/executor/executorworker.h:60]:
```suggestion
         * This method aims to fix the last modification date of a local file using the date stored in DB. This allows us to fix wrong
```

File: src/libsyncengine/propagation/executor/executorworker.h [hunk 5]
```diff
@@ -55,7 +55,16 @@ class ExecutorWorker : public OperationProcessor {
...
+         * EDIT operations generated on dehydrated placeholder.
```
Comment (luc-guyot-infomaniak) [inline src/libsyncengine/propagation/executor/executorworker.h:61]:
```suggestion
         * EDIT operations generated on dehydrated placeholders.
```

File: src/server/socketapi.cpp [hunk 6]
```diff
@@ -790,9 +792,9 @@ void SocketApi::command_GET_STRINGS(const QString &argument, SocketListener *lis
         {"COPY_PRIVATE_LINK_MENU_TITLE", tr("Copy private share link")},
     }};
     listener->sendMessage(QString("GET_STRINGS%1BEGIN").arg(MSG_CDE_SEPARATOR));
-    for (auto key_value : strings) {
-        if (argument.isEmpty() || argument == QString(key_value.first)) {
-            listener->sendMessage(QString("STRING%1%2%1%3").arg(MSG_CDE_SEPARATOR).arg(key_value.first, key_value.second));
+    for (auto &keyValue : strings) {
```
Comment (luc-guyot-infomaniak) [inline src/server/socketapi.cpp:795]:
```suggestion
    for (const auto &[key, value] : strings) {
```

File: test/libsyncengine/propagation/executor/testexecutorworker.cpp [hunk 7]
```diff
@@ -180,4 +162,42 @@ void TestExecutor::testCheckLiteSyncInfoForCreate() {
...
+        ofs << "abc";
+        ofs.close();
+    }
+
+    // Update DB
+    DbNode dbNode(0, 123, filename, filename, "lid", "rid", defaultTime, defaultTime, defaultTime, NodeTypeFile, defaultSize,
+                  "cs");
+    DbNodeId dbNodeId;
+    bool constraintError = false;
+    _syncPal->_syncDb->insertNode(dbNode, dbNodeId, constraintError);
```
Comment (luc-guyot-infomaniak) [inline test/libsyncengine/propagation/executor/testexecutorworker.cpp:182]: Public accessor?

File: test/libsyncengine/propagation/executor/testexecutorworker.cpp [hunk 8]
```diff
@@ -180,4 +162,42 @@ void TestExecutor::testCheckLiteSyncInfoForCreate() {
...
+    // Generate sync operation
+    std::shared_ptr<Node> node =
+        std::make_shared<Node>(std::nullopt, ReplicaSideLocal, filename, NodeTypeFile, "lid", defaultTime, 12345, defaultSize);
+    std::shared_ptr<Node> correspondingNode = std::make_shared<Node>(std::nullopt, ReplicaSideLocal, filename, NodeTypeFile,
+                                                                     "rid", defaultTime, defaultTime, defaultSize);
+    SyncOpPtr op = std::make_shared<SyncOperation>();
+    op->setAffectedNode(node);
+    op->setCorrespondingNode(correspondingNode);
+
+    _syncPal->_executorWorker->fixModificationDate(op, path);
```
Comment (luc-guyot-infomaniak) [inline test/libsyncengine/propagation/executor/testexecutorworker.cpp:193]: Public accessor?

Comment (ClementKunz) [reply test/libsyncengine/propagation/executor/testexecutorworker.cpp:193]: Not sure if it is a good idea to have a public accessor on `_executorWorker `. We do not want any other module to be able to access and modify the `Executor` (except from the test class).
What do you think?
IMHO, we definitively put to too much friend classes, but it is ok to define the test class as a friend of the base class (e.g. : `TestExecutorWorker` is a friend of `ExecutorWorker`). However, the SyncPal class has too many friends, and each unit test class should depend only on the class it is testing (e.g. : `TestExecutorWorker` should not have to allocate a SyncPal variable). But such a big change should be the topic of a dedicated PR

File: src/server/socketapi.cpp (hunk 6 above)
Comment (ClementKunz) [reply src/server/socketapi.cpp]: Done

File: test/libsyncengine/propagation/executor/testexecutorworker.cpp (hunk 7 above)
Comment (ClementKunz) [reply test/libsyncengine/propagation/executor/testexecutorworker.cpp]: Done

**Review by luc-guyot-infomaniak — APPROVED** [review]

**Pull request merged**